BuilderBuilder.Test/AssertHelper.cs

```diff
@@ -1,86 +1,73 @@
-            ListEq(result, compareBy, expected.AsEnumerable());
-        }
-        public static void ListEq<T>(IEnumerable<T> result, params T[] expected) {
-            ListEq(result, expected.AsEnumerable());
-        }
+    private static void ListEq<T, TC>(IEnumerable<T> result, Func<T, TC> compareBy, IEnumerable<T> expected) {
+        ListEq(result.Select(compareBy), expected.Select(compareBy));
+    }
 
-        public static void ListEq<T, C>(IEnumerable<T> result, Func<T, C> compareBy, IEnumerable<T> expected) {
-            ListEq(result.Select(compareBy), expected.Select(compareBy));
-        }
-        public static void ListEq<T, C>(IEnumerable<T> result, Func<T, C> compareBy, IEnumerable<C> expected) {
-            ListEq(result.Select(compareBy), expected);
+    private static void ListEq<T>(IEnumerable<T> result, IEnumerable<T> expected) {
+        var actualArr = result as T[] ?? result.ToArray();
+        var expectedArr = expected as T[] ?? expected.ToArray();
+        var shouldHaves = Missing(actualArr, expectedArr);
+        var shouldntHaves = new List<T>();
+        if (actualArr.Length + shouldHaves.Count != expectedArr.Length) {
+            shouldntHaves = Missing(expectedArr, actualArr);
         }
-        public static void ListEq<T>(IEnumerable<T> result, IEnumerable<T> expected) {
-            List<T> shouldHaves = Missing(result, expected);
-            List<T> shouldntHaves = new List<T>();
-            if (result.Count() + shouldHaves.Count != expected.Count()) {
-                shouldntHaves = Missing(expected, result);
-            }
 
-            if (shouldHaves.Count > 0 || shouldntHaves.Count > 0) {
-                Assert.Fail(ErrorList("Should haves", shouldHaves) + ". " + ErrorList("Shouldn't haves", shouldntHaves));
-            }
+        if (shouldHaves.Count > 0 || shouldntHaves.Count > 0) {
+            Assert.Fail(ErrorList("Should haves", shouldHaves) + ". " + ErrorList("Shouldn't haves", shouldntHaves));
         }
+    }
 
-        private static List<T> Missing<T>(IEnumerable<T> checkList, IEnumerable<T> source) {
-            HashSet<T> sourceHashSet = new HashSet<T>(checkList);
-            return source.Where(b => !sourceHashSet.Contains(b)).ToList();
-        }
+    private static List<T> Missing<T>(IEnumerable<T> checkList, IEnumerable<T> source) {
+        var sourceHashSet = new HashSet<T>(checkList);
+        return source.Where(b => !sourceHashSet.Contains(b)).ToList();
+    }
 
-        private static string ErrorList<T>(string description, IEnumerable<T> list) {
-            if (!list.Any()) {
-                return "";
-            }
-            var errors = string.Join(", ", list.Select(i => i.ToString()));
-            return $"{description}: {errors}";
+    private static string ErrorList<T>(string description, IEnumerable<T> list) {
+        var arr = list as T[] ?? list.ToArray();
+        if (!arr.Any()) {
```

Comment:
On an array it is a constant operation, but this is an IEnumerable so it has to start enumerating before it can know that there is any value. I changed the parameters of the private methods to IList so there is no chance for multiple enumerations.